File: packages/unity_bridge_contracts/lib/src/bridge_envelope.dart
import 'package:unity_bridge_contracts/src/bridge_constants.dart';
import 'package:unity_bridge_contracts/src/bridge_errors.dart';
import 'package:unity_bridge_contracts/src/bridge_message_type.dart';

/// Versioned JSON envelope shared by Flutter and Unity.
final class BridgeEnvelope {
  const BridgeEnvelope({
    required this.schemaVersion,
    required this.messageId,
    required this.sessionId,
    required this.attemptId,
    required this.levelDefinitionId,
    required this.revision,
    required this.type,
    required this.payload,
    this.requestId,
  });

  final int schemaVersion;
  final String messageId;
  final String sessionId;
  final String attemptId;
  final String levelDefinitionId;
  final int revision;
  final BridgeMessageType type;
  final String? requestId;
  final Map<String, Object?> payload;

  Map<String, Object?> toJson() => {
    'schemaVersion': schemaVersion,
    'messageId': messageId,
    'sessionId': sessionId,
    'attemptId': attemptId,
    'levelDefinitionId': levelDefinitionId,
    'revision': revision,
    'type': type.wireName,
    if (requestId != null) 'requestId': requestId,
    'payload': payload,
  };

  static BridgeEnvelope fromJson(Map<String, Object?> json) {
    final schemaVersion = json['schemaVersion'];
    if (schemaVersion is! int) {
      throw MissingRequiredFieldError('schemaVersion');
    }
    if (schemaVersion != kBridgeSchemaVersion) {
      throw UnknownSchemaVersionError(schemaVersion);
    }

    String requireString(String key) {
      final value = json[key];
      if (value is! String || value.isEmpty) {
        throw MissingRequiredFieldError(key);
      }
      return value;
    }

    final revision = json['revision'];
    if (revision is! int) {
      throw MissingRequiredFieldError('revision');
    }

    final typeRaw = json['type'];
    if (typeRaw is! String || typeRaw.isEmpty) {
      throw MissingRequiredFieldError('type');
    }

    late final BridgeMessageType type;
    try {
      type = BridgeMessageType.parse(typeRaw);
    } on FormatException {
      throw UnknownMessageTypeError(typeRaw);
    }

    final payloadRaw = json['payload'];
    if (payloadRaw is! Map) {
      throw MissingRequiredFieldError('payload');
    }

    final requestId = json['requestId'];
    return BridgeEnvelope(
      schemaVersion: schemaVersion,
      messageId: requireString('messageId'),
      sessionId: requireString('sessionId'),
      attemptId: requireString('attemptId'),
      levelDefinitionId: requireString('levelDefinitionId'),
      revision: revision,
      type: type,
      requestId: requestId is String && requestId.isNotEmpty ? requestId : null,
      payload: Map<String, Object?>.from(payloadRaw),
    );
  }
}

File: unity/ArabSolitaireUnity/Assets/ArabSolitaire/Bridge/BridgeEnvelope.cs
using System;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;

namespace ArabSolitaire.Bridge
{
    [Serializable]
    public sealed class BridgeEnvelope
    {
        public int schemaVersion;
        public string messageId = string.Empty;
        public string sessionId = string.Empty;
        public string attemptId = string.Empty;
        public string levelDefinitionId = string.Empty;
        public int revision;
        public string type = string.Empty;
        public string requestId;
        public JObject payload = new();

        public static BridgeEnvelope FromJson(string json)
        {
            var envelope = JsonConvert.DeserializeObject<BridgeEnvelope>(json)
                ?? throw new MissingRequiredFieldError("root");
            Validate(envelope);
            return envelope;
        }

        public static BridgeEnvelope FromJObject(JObject json)
        {
            var envelope = json.ToObject<BridgeEnvelope>()
                ?? throw new MissingRequiredFieldError("root");
            Validate(envelope);
            return envelope;
        }

        public static void Validate(BridgeEnvelope envelope)
        {
            if (envelope.schemaVersion != BridgeConstants.SchemaVersion)
            {
                throw new UnknownSchemaVersionError(envelope.schemaVersion);
            }

            Require(envelope.messageId, nameof(messageId));
            Require(envelope.sessionId, nameof(sessionId));
            Require(envelope.attemptId, nameof(attemptId));
            Require(envelope.levelDefinitionId, nameof(levelDefinitionId));
            Require(envelope.type, nameof(type));

            if (!BridgeMessageTypeExtensions.TryParse(envelope.type, out _))
            {
                throw new UnknownMessageTypeError(envelope.type);
            }

            if (envelope.payload == null)
            {
                throw new MissingRequiredFieldError(nameof(payload));
            }
        }

        public string ToJson() => JsonConvert.SerializeObject(this);

        private static void Require(string value, string field)
        {
            if (string.IsNullOrWhiteSpace(value))
            {
                throw new MissingRequiredFieldError(field);
            }
        }
    }

    [Serializable]
    public sealed class TransitionResultPayload
    {
        public bool accepted;
        public JObject nextState;
        public int moveCost;
        public JArray events = new();
        public int newRevision;
        public string rejectionReason;
    }
}
